src/compiler/transformers/classFields.ts

```diff
@@ -506,6 +786,165 @@ namespace ts {
+            return createPropertyAccess(
+                createObjectLiteral([
+                    createSetAccessor(
+                        /*decorators*/ undefined,
+                        /*modifiers*/ undefined,
+                        "value",
+                        [createParameter(
+                            /*decorators*/ undefined,
+                            /*modifiers*/ undefined,
+                            /*dotDotDotToken*/ undefined, "x",
+                            /*questionToken*/ undefined,
+                            /*type*/ undefined,
+                            /*initializer*/ undefined
+                        )],
+                        createBlock(
+                            [createExpressionStatement(
+                                visitNode(
+                                    createAssignment(node, createIdentifier("x")),
+                                    visitor
+                                )
+                            )]
+                        )
+                    )
+                ]),
+                "value"
+            );
+        }
+
+        function transformDestructuringAssignmentTarget(node: ArrayLiteralExpression | ObjectLiteralExpression) {
+            const hasPrivateIdentifiers = isArrayLiteralExpression(node) ?
+                forEach(node.elements, isPrivateIdentifierPropertyAccessExpression) :
+                forEach(node.properties, property => isPropertyAssignment(property) && isPrivateIdentifierPropertyAccessExpression(property.initializer));
+            if (!hasPrivateIdentifiers) {
+                return node;
+            }
+            if (isArrayLiteralExpression(node)) {
+                // Transforms private names in destructuring assignment array bindings.
+                //
+                // Source:
+                // ([ this.#myProp ] = [ "hello" ]);
+                //
+                // Transformation:
+                // [ { set value(x) { this.#myProp = x; } }.value ] = [ "hello" ];
+                return updateArrayLiteral(
+                    node,
+                    node.elements.map(
+                        expr => isPrivateIdentifierPropertyAccessExpression(expr) ?
+                            wrapPrivateIdentifierForDestructuringTarget(expr) :
+                            expr
+                    )
+                );
+            }
+            else {
+                // Transforms private names in destructuring assignment object bindings.
+                //
+                // Source:
+                // ({ stringProperty: this.#myProp } = { stringProperty: "hello" });
+                //
+                // Transformation:
+                // ({ stringProperty: { set value(x) { this.#myProp = x; } }.value }) = { stringProperty: "hello" };
+                return updateObjectLiteral(
+                    node,
+                    node.properties.map(

```

Comment:
```suggestion
                    visitNodes(node.properties, visitObjectAssignmentTarget, isObjectLiteralElementLike)
```

Where `visitObjectAssignmentTarget` is essentially the arrow function below, except you should call `visitor` on `prop` when its not a property assignment to a `PrivateIdentifierPropertyAccessExpression`.

Also, you are not handling the case for defaults in the initializer. You should be using `isPrivateIdentifierPropertyAccessExpression(getTargetOfBindingOrAssignmentElement(prop))`. Please add a test to verify this case:

```ts
({ a: x.#a = 1 } = {});
```

Then you won't need the guard at the top of the function.